### PR TITLE
MIDI-92: Improvements to Discriminator and Data Normalization

### PR DESCRIPTION
**Pull request by SamuelJanas**
This PR summarizes the successful approaches and the ones that didn't yield significant benefits.

## What worked:
- **Data Normalization**: Disallowed the discriminator to judge the outcome based on different input ranges.
- **Residual Blocks**: Made the discriminator more context-aware.

## What didn't:
- **Better Generator**: Improving the generator didn't influence the outcomes significantly. However, it might aid with stability now after modifying the discriminator model.
- **Hyperparameters Tuning**: After data normalization, hyperparameters had a diminished role compared to the earlier version.

Related Issues:
[MIDI-92](https://sacrebleu.atlassian.net/browse/MIDI-92)
[MIDI-98](https://sacrebleu.atlassian.net/browse/MIDI-98)

[MIDI-92]: https://sacrebleu.atlassian.net/browse/MIDI-92?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[MIDI-98]: https://sacrebleu.atlassian.net/browse/MIDI-98?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by SamuelJanas** [comment]
After reviewing papers on DCGANs applied to 1D data, most, like the [Arbitrary Signal Generator](https://ieeexplore.ieee.org/document/8909059), adopt a similar model architecture to ours, differing mainly in model size and hyperparameters, especially 'nz' (noise_size). This suggests we either need to optimize our hyperparameters specifically for our use case or explore alternative GAN architectures better suited for multichannel 1D data, although the former seems more plausible due to limited research in the latter.

**Comment by SamuelJanas** [comment]
To get to a point at which I can judge if a run is good or not faster, I changed the batch size: ~64~ -> 16. GANs don't really need test/train splits, now they are all used for training.
I'm planning to list experiments I've tried with some description to understand failing conditions better:
https://docs.google.com/spreadsheets/d/1bE3VIL-rMds4F5k2fm_IGL3w4Wfjp7h3tDbrbNOpCWw/edit?usp=sharing
google docs isn't the best solution, it's an easy one but I'll be grateful for different suggestions

**Comment by roszcz** [comment]
I think a spreadsheet can work very well for this :)

**Comment by SamuelJanas** [comment]
https://wandb.ai/sjanas/ECG%20GAN/runs/vghvhxc4
As a break from brainstorming for ideas to "fix" the GAN, I ran our baseline model on 2 channels. I'm surprised that the generations towards the latter part of training look much worse than ~100 batches in. Not sure yet if we get anything out of this but it's interesting nevertheless.

**Comment by roszcz** [comment]
My guess would be that this indicates one of the models learning faster than the other - do you have learning rate manipulations on your list of things to try? :)

**Comment by SamuelJanas** [comment]
> My guess would be that this indicates one of the models learning faster than the other - do you have learning rate manipulations on your list of things to try? :)

One run is already there on the list, added some notes to runs, that one will have to be run overnight as I can't really tell which way it's going 😅

**Comment by SamuelJanas** [comment]
Currently the gradient vanishes instantly for TGAN, pushing it here for potential future work but it can be ignored for now. Switching my focus to VQGAN.

**Comment by SamuelJanas** [comment]
interpolation
Amazingly enough, interpolations between two noises seem like a fluid transition, feel free to experiment with the seed or `number_of_interpolations` to get different results. 
If you don't have a checkpoint stored locally, it should automatically download from huggingface.

To run use `python -m evals.interpolate_ecg`

**Comment by SamuelJanas** [comment]
After closer inspection on our model and finding out that there's a need to standardize the inputs to the discriminator I have removed `tanh()` from the generator as it's an easy workaround. This had a surprising impact on the training process.
 - **Previously with tanh**: The D_g_z2 value, representing the generator's ability to fool the discriminator, consistently dropped to around 0.05 and stayed there [See this trend](https://wandb.ai/sjanas/ECG%20GAN/runs/n0vxsror).
image


- **Post-removal of tanh**: We're seeing a change. Not only is the D_g_z2 value no longer plummeting, but our discriminator is also now in a balanced position, ~often~ unsure if an input is real or generated [Check out the updated run](https://wandb.ai/sjanas/ECG%20GAN/runs/al740b6n).
image


What that means: We could improve our discriminator to potentially force the generator to come up with better generations, stability greatly improved.

**Comment by SamuelJanas** [comment]
Since tanh is a rather important activation in DCGAN models, I found out that normalizing the data and introducing the tanh again had a similar effect on generator. [wandb](https://wandb.ai/sjanas/ECG%20GAN/runs/5chghhke=)


There are some nuances, mainly since the data is normalized we introduced some noise to generations, this unfortunately persists after denormalization. Here's a gif after denorm:


Here are some statistics from `data_showcase.ipynb`

For future we could either clean the dataset some more removing batches with extreme values as seen in the chart or apply some smoothing, that could work just as well.

I'll try training the model with the second channel as the data looks more 'organized'. The second step would be to improve the discriminator as mentioned in previous comment.

**Comment by SamuelJanas** [comment]
I introduced ResidualBlocks to discriminator to make it more context aware with great success! [wandb](https://wandb.ai/sjanas/ECG%20GAN/runs/3gg9i73h)
sample generation after one epoch of training:

This introduced two issues however. Performance took a hit as the model is more complex right now and the stability was greatly reduced. I believe we can overcome the second issue with the right hyperparameters and/or longer training
Overall, big improvements!


EDIT: We can see [here](https://wandb.ai/sjanas/ECG%20GAN/runs/fz2ptaxy) that the generator is falling behind just slightly to the discriminator even after tweaking the learning rate. That being said I think that the results are good enough, there is clear rhythm and the shapes look the way they're supposed to.

File: utils/dataloaders.py [hunk 1]
```diff
@@ -53,8 +67,8 @@ def seed_worker(worker_id):
 def create_dataloader(
     cfg: DictConfig, seed: int = None, splits=["train", "validation", "test"]
 ) -> (DataLoader, DataLoader, DataLoader):
-    # Combine the three datasets
-    datasets = [load_dataset("roszcz/ecg-segmentation-ltafdb", split=split) for split in splits]
+    # Use cleaned up dataset ~3% less samples, but less noise
+    datasets = [load_dataset("SneakyInsect/ltafdb_preprocessed", split=split) for split in splits]
```
Comment (roszcz) [inline utils/dataloaders.py:71]: Can you also include the code used to pre-process the dataset?

Comment (SamuelJanas) [reply utils/dataloaders.py:71]: it's here now, I found a workaround using your mask dataset 😄

File: utils/preprocess-ltafdb.py
Comment (roszcz) [inline utils/preprocess-ltafdb.py:1]: it's not recommended to use dashes in python file names, `preprocess_ltafdb.py` will be much better

**Review by roszcz — APPROVED** [review]

File: utils/preprocess-ltafdb.py
Comment (SamuelJanas) [reply utils/preprocess-ltafdb.py:1]: sure! that was a typo
